// readme-aura/dist/parser.js
import { readFile } from 'node:fs/promises';
import { resolve, relative, dirname } from 'node:path';
import { unified } from 'unified';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import { visit } from 'unist-util-visit';
import { parseMeta } from './renderer.js';
const AURA_LANG = 'aura';
export async function parseSource(sourcePath, assetsDir = '.github/assets', outputPath = 'README.md') {
    const source = await readFile(sourcePath, 'utf-8');
    const blocks = [];
    let blockIndex = 0;
    // Compute asset path relative to the output file
    const absAssets = resolve(assetsDir);
    const absOutput = resolve(outputPath);
    const relAssets = './' + relative(dirname(absOutput), absAssets).replace(/\\/g, '/');
    const file = await unified()
        .use(remarkParse)
        .use(() => (tree) => {
        visit(tree, 'code', (node, index, parent) => {
            const codeNode = node;
            if (codeNode.lang !== AURA_LANG || index === undefined || !parent)
                return;
            blocks.push({
                index: blockIndex,
                content: codeNode.value,
                meta: codeNode.meta ?? undefined,
            });
            const { link, inline } = parseMeta(codeNode.meta ?? undefined);
            if (inline) {
                const htmlNode = {
                    type: 'html',
                    value: `<!-- __AURA_INLINE_${blockIndex}__ -->`,
                };
                parent.children[index] = htmlNode;
            }
            else {
                const imageNode = {
                    type: 'image',
                    url: `${relAssets}/readme-aura-component-${blockIndex}.svg`,
                    alt: `readme-aura-component-${blockIndex}`,
                    title: 'Generated by readme-aura',
                };
                const paragraphChild = link
                    ? { type: 'link', url: link, children: [imageNode] }
                    : imageNode;
                const paragraphNode = {
                    type: 'paragraph',
                    children: [paragraphChild],
                };
                parent.children[index] = paragraphNode;
            }
            blockIndex++;
        });
    })
        .use(remarkStringify)
        .process(source);
    // Merge consecutive inline placeholders: remove blank lines between them
    // so they end up on one line and GitHub renders them as inline elements.
    let md = String(file);
    let prev;
    do {
        prev = md;
        md = md.replace(/(<!-- __AURA_INLINE_\d+__ -->)\n\n(<!-- __AURA_INLINE_\d+__ -->)/g, '$1$2');
    } while (md !== prev);
    return {
        markdown: md,
        blocks,
        relAssets,
    };
}
