// readme-aura/dist/renderer.js
import satori from 'satori';
import { transform } from 'sucrase';
import { readFile } from 'node:fs/promises';
import { resolve, extname, isAbsolute } from 'node:path';
export function parseMeta(meta) {
    const defaults = { width: 800, height: 400 };
    if (!meta)
        return defaults;
    const pairs = [...meta.matchAll(/(\w+)=(\d+)/g)];
    for (const [, key, value] of pairs) {
        if (key === 'width')
            defaults.width = parseInt(value, 10);
        if (key === 'height')
            defaults.height = parseInt(value, 10);
    }
    const linkMatch = meta.match(/link=["']([^"']+)["']/);
    if (linkMatch) {
        defaults.link = linkMatch[1];
    }
    if (/\binline\b/.test(meta)) {
        defaults.inline = true;
    }
    const alignMatch = meta.match(/\balign=(center|left|right)\b/);
    if (alignMatch) {
        defaults.align = alignMatch[1];
    }
    return defaults;
}
export function createElement(type, props, ...children) {
    const flat = children.flat().filter((c) => c != null && c !== false && c !== true);
    return {
        type,
        props: {
            ...(props ?? {}),
            children: flat.length === 0 ? undefined : flat.length === 1 ? flat[0] : flat,
        },
    };
}
export function transpileJsx(jsxString, context) {
    const wrapped = `return (${jsxString.trim()});`;
    const { code } = transform(wrapped, {
        transforms: ['jsx'],
        jsxRuntime: 'classic',
        production: true,
    });
    const argNames = ['React', ...(context ? Object.keys(context) : [])];
    const argValues = [{ createElement }, ...(context ? Object.values(context) : [])];
    const fn = new Function(...argNames, code);
    const element = fn(...argValues);
    if (!element || typeof element !== 'object' || !('type' in element)) {
        throw new Error('JSX did not produce a valid element. Make sure your code returns a single root element.');
    }
    return element;
}
function extractStyles(node, styles) {
    if (!node || typeof node !== 'object')
        return node;
    const el = node;
    if (el.type === 'style') {
        if (el.props?.children) {
            styles.push(String(el.props.children));
        }
        return null;
    }
    if (el.props?.children) {
        if (Array.isArray(el.props.children)) {
            el.props.children = el.props.children
                .map((c) => extractStyles(c, styles))
                .filter((c) => c !== null);
        }
        else {
            const processed = extractStyles(el.props.children, styles);
            if (processed === null) {
                delete el.props.children;
            }
            else {
                el.props.children = processed;
            }
        }
    }
    return el;
}
const MIME_TYPES = {
    png: 'image/png',
    jpg: 'image/jpeg',
    jpeg: 'image/jpeg',
    gif: 'image/gif',
    webp: 'image/webp',
    svg: 'image/svg+xml',
};
export async function resolveLocalImages(element, basePath) {
    if (!element || typeof element !== 'object')
        return element;
    const el = element;
    if (el.type === 'img' && typeof el.props?.src === 'string') {
        const src = el.props.src;
        if (!src.startsWith('http://') && !src.startsWith('https://') && !src.startsWith('data:')) {
            const fullPath = isAbsolute(src) ? src : resolve(basePath, src);
            const data = await readFile(fullPath);
            const ext = extname(fullPath).slice(1).toLowerCase();
            const mime = MIME_TYPES[ext] ?? `image/${ext}`;
            el.props.src = `data:${mime};base64,${data.toString('base64')}`;
        }
    }
    if (el.props?.children) {
        if (Array.isArray(el.props.children)) {
            el.props.children = (await Promise.all(el.props.children.map((c) => resolveLocalImages(c, basePath))));
        }
        else if (typeof el.props.children === 'object') {
            el.props.children = (await resolveLocalImages(el.props.children, basePath));
        }
    }
    return el;
}
export async function renderBlock(block, fonts, context, basePath) {
    const { width, height } = parseMeta(block.meta);
    let element;
    try {
        element = transpileJsx(block.content, context);
    }
    catch (err) {
        throw new Error(`Failed to transpile JSX in block ${block.index}: ${err.message}`, {
            cause: err,
        });
    }
    if (basePath) {
        element = await resolveLocalImages(element, basePath);
    }
    const extractedStyles = [];
    element = extractStyles(element, extractedStyles);
    try {
        let svg = await satori(element, {
            width,
            height,
            fonts,
            loadAdditionalAsset: async (code, segment) => {
                if (code === 'emoji') {
                    const codepoint = [...segment].map((c) => (c.codePointAt(0) ?? 0).toString(16)).join('-');
                    const url = `https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/${codepoint}.svg`;
                    const res = await fetch(url);
                    const svgText = await res.text();
                    return `data:image/svg+xml;base64,${Buffer.from(svgText).toString('base64')}`;
                }
                return '';
            },
        });
        // Unpack nested SVGs encoded by Satori as data:image/svg+xml;utf8,... <image> tags
        svg = svg.replace(/<image\s+([^>]*?)href="data:image\/svg\+xml;utf8,([^"]+)"([^>]*?)\/>/g, (match, before, encoded, after) => {
            try {
                const decoded = decodeURIComponent(encoded);
                const contentMatch = decoded.match(/<svg[^>]*>([\s\S]*?)<\/svg>/i);
                const content = contentMatch ? contentMatch[1] : decoded;
                return `<svg ${before.trim()} ${after.trim()}>${content}</svg>`;
            }
            catch {
                return match;
            }
        });
        if (extractedStyles.length > 0) {
            const combinedStyles = extractedStyles.join('\n');
            svg = svg.replace('</svg>', `<style>\n${combinedStyles}\n</style>\n</svg>`);
        }
        svg = '<!-- Generated by readme-aura https://github.com/collectioneur/readme-aura -->\n' + svg;
        return svg;
    }
    catch (err) {
        throw new Error(`Failed to render block ${block.index} to SVG: ${err.message}`, {
            cause: err,
        });
    }
}
